lib/r10k/module_loader/puppetfile.rb

```
@@ -0,0 +1,190 @@
+          desired_contents: @desired_contents,
+          purge_exclusions: @purge_exclusions
+        }.freeze
+
+      rescue SyntaxError, LoadError, ArgumentError, NameError => e
+        raise R10K::Error.wrap(e, _("Failed to evaluate %{path}") % {path: @puppetfile})
+      end
+
+
+      ##
+      ## set_forge, set_moduledir, and add_module are used directly by the DSL class
+      ##
+
+      # @param [String] forge
+      def set_forge(forge)
+        @forge = forge
+      end
+
+      # @param [String] moduledir
+      def set_moduledir(moduledir)
+        @moduledir = if Pathname.new(moduledir).absolute?
+          moduledir
+        else
+          File.join(@basedir, moduledir)
+        end
+      end
+
+      # @param [String] name
+      # @param [Hash, String, Symbol] args Calling with anything but a Hash is
+      #   deprecated. The DSL will now convert String and Symbol versions to
+      #   Hashes of the shape
+      #     { version: <String or Symbol> }
+      #
+      def add_module(name, args)
+        if !args.is_a?(Hash)
+          args = { version: args }
+        end
+
+        args[:overrides] = @overrides
+
+        if install_path = args.delete(:install_path)
+          install_path = resolve_install_path(install_path)
+          validate_install_path!(install_path, name)
+        else
+          install_path = @moduledir
+        end
+
+        if @default_branch_override != nil
+          args[:default_branch_override] = @default_branch_override
+        end
+
+        mod = R10K::Module.new(name, install_path, args, @environment)
+        mod.origin = :puppetfile
+
+        # Do not load modules if they would conflict with the attached
+        # environment
+        if @environment && @environment.module_conflicts?(mod)
+          mod = nil
+          return @modules
+        end
+
+        @modules << mod
+      end
+
+     private
+
+      # @param [Array<R10K::Module>] modules
+      def validate_no_duplicate_names!(modules)
```

Why does this have a `!`? It doesn't modify `modules` at all, does it?

---

This is a clojurism that doesn't mean the same thing in Ruby, agree it should be dropped.

---

A method that ends in a bang means that the method is more dangerous or surprising than one would expect. Typically in core classes it is used for methods that mutate the receiver. It is also used for methods that raise on failure instead of return data, or methods that may not be chainable in ways that similar methods are. Bang methods are always paired with a non-bang version. https://stackoverflow.com/questions/612189/why-are-exclamation-marks-used-in-ruby-methods

In Clojure the bang means a method should not be used in a transaction - I think we use it especially in file-sync for things that modify the file system (though I wouldn't be surprised if we use it in functions that raise as well). https://stackoverflow.com/questions/20606249/when-to-use-exclamation-mark-in-clojure-or-lisp

The exclamation mark is there because its a validation function that raises rather than returns data, which I think can be an accurate usage in Ruby, but it isn't paired with a non-bang version, so I agree it should be dropped.